subprojects/core/src/integTest/groovy/org/gradle/api/invocation/GradleLifecycleBeforeProjectEagerExecutionIntegrationTest.groovy

```diff
@@ -0,0 +1,245 @@
+                println "lifecycle.beforeProject: \${name}"
+            }
+            include(":a")
+        """
+        file("a/build.gradle") << ""
+        buildFile """
+            project(":a") {
+                println("before")
+                it.$mutableStateAccess
+                println("after")
+            }
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains "before\nlifecycle.beforeProject: a\nafter"
+
+        where:
+        mutableStateAccess << [
+            "buildDir",
+            "group",
+            "version",
+            "description",
+            "status",
+            "evaluate()",
+            "beforeEvaluate{}",
+            "afterEvaluate{}",
+            "evaluationDependsOn(':a')",
+            "evaluationDependsOnChildren()",
+            "bindAllModelRules()",
+            "prepareForRuleBasedPlugins()",
+            "services",
+            "serviceRegistryFactory",
+            "configurationActions",
+            "modelRegistry",
+            "model",
+            "fireDeferredConfiguration()",
+            "addDeferredConfiguration{}",
+            "normalization",
+            "normalization{}",
+            "dependencyLocking",
+            "dependencyLocking{}",
+            "tasks",
+            "getAllTasks(true)",
+            "getTasksByName('foo', true)",
+            "task('foo')",
+            "defaultTasks",
+            "getAnt()",
+            "ant{}",
+            "createAntBuilder()",
+            "layout",
+            "repositories",
+            "dependencies",
+            "state",
+            "extensions",
+            "buildscript",
+            "configurations",
+            "components",
+            "artifacts",
+            "convention",
+            "plugins",
+            "pluginManager",
+            "getProperties()",
+            "findProperty('foo')",
+            "getProperties()",
+            "hasProperty('foo')",
+            "apply{}"
+        ]
+    }
+
+    def 'lifecycle.beforeProject is executed only once for a project'() {
+        given:
+        settingsFile """
+            rootProject.name = 'root'
+            gradle.lifecycle.beforeProject {
+                println "lifecycle.beforeProject for \${name}"
+            }
+            include(":a")
+        """
+
+        file("a/build.gradle") << ""
+
+        buildFile """
+            allprojects { $projectMutableStateAccess }
+            $secondEagerExecutionBlock
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        output.count("lifecycle.beforeProject for root") == 1
+        output.count("lifecycle.beforeProject for a") == 1
+
+        where:
+        secondEagerExecutionBlock << [
+            "subprojects { $projectMutableStateAccess }",
+            "afterEvaluate { allprojects { $projectMutableStateAccess } }"
+        ]
+    }
+
+    def 'lifecycle.beforeProject can be executed before project.#api'() {
```

Comment:
❓  Is it _executed before_ or is it triggered _from inside_ the outer callback when the mutable state access happens?

I think adding a `println` before the mutable state access inside `invocation` lambda would make it clear

Reply:
I think the it's `triggered from inside the outer callback` is better wording here. However, I didn't understand what exactly will become cleaner with print before mutable state access?
It would make sense if `lifecycle.beforeProject` been doing a print as well, but it doesn't.

Reply:
Nothing prevents us from adding the println there as well.

For me, the expected output in the end of the test should clearly show which callback starts running at what time relative to the other callback and mutable access. I think adding more println statements is a reasonable way of achieving this.

Reply:
I see your point. Mine is that having "real" access scenario (register a property and then getting the property) AND printing in callback are serving to same purpose. First is a bit implicit, but "real", the second is explicit in a sense of printing(it's better shows up callbacks execution order, I'm agree), but "just" printing.
I don't think that having both approaches make any sense - it inflates an expected output for the test.

I don't have any strong stance in this regard, I just think we shouldn't use both approaches simultaniosly

Reply:
Added additional prints while keeping property access as well. The expected output is still more or less readable.

Reply:
I find the new updated tests easier to understand, because they now highlight the order in which the callbacks _have started_, and that once callback _triggers_ the other. I think the tests should be as explicit as possible, and now they are 👍